src/transformers/models/idefics3/processing_idefics3.py

```diff
@@ -0,0 +1,351 @@
+# coding=utf-8
+# Copyright 2024 The HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Processor class for Idefics3.
+"""
+
+from typing import TYPE_CHECKING, List, Optional, Union
+
+from ...feature_extraction_utils import BatchFeature
+from ...image_utils import ImageInput, is_valid_image, load_image
+from ...processing_utils import ProcessorMixin
+from ...tokenization_utils_base import AddedToken, BatchEncoding, PaddingStrategy, TextInput, TruncationStrategy
+from ...utils import TensorType, logging
+
+
+if TYPE_CHECKING:
+    from ...tokenization_utils_base import PreTokenizedInput
+
+
+logger = logging.get_logger(__name__)
+
+
+def is_url(val) -> bool:
+    return isinstance(val, str) and val.startswith("http")
+
+
+def is_image_or_image_url(elem):
+    return is_url(elem) or is_valid_image(elem)
+
+
+def _prompt_split_image(image_seq_len, image_rows, image_cols, fake_token_around_image, image_token):
+    """Prompt with expanded image tokens for when the image is split into patches."""
+    text_splitted_images = ""
+    for n_h in range(image_rows):
+        for n_w in range(image_cols):
+            text_splitted_images += (
+                f"{fake_token_around_image}"
+                + f"<row_{n_h + 1}_col_{n_w + 1}>"
+                + f"{image_token}" * image_seq_len
+            )
+        text_splitted_images += "\n"
+
+    text_splitted_images += (
+        f"\n{fake_token_around_image}"
+        + "<global-img>"
+        + f"{image_token}" * image_seq_len
+        + f"{fake_token_around_image}"
+    )
+    return text_splitted_images
+
+
+def _prompt_single_image(image_seq_len, fake_token_around_image, image_token):
+    """Prompt with expanded image tokens for a single image."""
+    return f"{fake_token_around_image}" + "<global-img>" + f"{image_token}" * image_seq_len + f"{fake_token_around_image}"
```

Comment:
btw, in Chameleon, we use `image_start_token` & `image_end_token` (and `boi_token_id` & `eoi_token_id` for the token ids)

Reply:
Yeah, I guess we can try to unify these special tokens after adding a MultiModalTokenizer wrapper :)

Reply:
Let's!